File: blazor-server/Models/Assignment.cs
namespace Projects.Models
{
  public class Assignment : BaseEntity
  {
    public int AssignmentId { get; set; }
    public required string Title { get; set; } = string.Empty;
    public string? Description { get; set; } = string.Empty;
    public required int ProjectId { get; set; }
    public Project? Project { get; set; }
    public required int StaffId { get; set; }
    public Staff? Staff { get; set; }
    public List<Expense>? Expenses { get; set; } = [];
    public List<Gain>? Gains { get; set; } = [];
    public required DateTime StartDate { get; set; } = DateTime.UtcNow;
    public DateTime? EndDate { get; set; }
    public required AssignmentStatus Status { get; set; } = AssignmentStatus.ACTIVE;
  }

  public enum AssignmentStatus
  {
    ACTIVE,
    INACTIVE,
    COMPLETED
  }
}

File: go-gin/ent/schema/assignment.go
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/mixin"
	"github.com/google/uuid"
)

// Assignment holds the schema definition for the Assignment entity.
type Assignment struct {
	ent.Schema
}

// Fields of the Assignment.
func (Assignment) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).
			Default(uuid.New).
			StorageKey("oid"),

		field.String("title").
			NotEmpty(),

		field.String("description").
			Optional().
			Nillable(),

		field.Time("start_date"),

		field.Time("end_date").
			Optional().
			Nillable(),

		field.Enum("status").
			Values("active", "in_active", "completed"),
	}
}

func (Assignment) Mixin() []ent.Mixin {
	return []ent.Mixin{
		mixin.Time{},
		// Or, mixin.CreateTime only for create_time
		// and mixin.UpdateTime only for update_time.
	}
}

// Edges of the Assignment.
func (Assignment) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("user", User.Type).
			Ref("assignments").
			Unique().
			Required().
			Immutable().
			Comment("An assignment only has one user and can't be changed after creation"),

		edge.From("project", Project.Type).
			Ref("assignments").
			Unique().
			Required().
			Immutable().
			Comment("An assignment only has one project and can't be changed after creation"),

		edge.To("statements", Statement.Type).
			Comment("An assignment can have many statements."),
	}
}
